pandas/core/algorithms.py

```diff
@@ -1255,6 +1259,18 @@ def compute(self, method: str) -> Series:
 
         dropped = self.obj.dropna()
 
+        if is_extension_array_dtype(dropped.dtype):
```

Comment:
why is this prefereable to hanlding on L1281 with the other dtypes?

Reply:
bc ensure_data does the wrong thing with MaskedArrays, np.asarray(obj) defaults to object dtype.

we _could_ kludge _ensure_data to work in cases where we dont have any pd.NAs, but doing it here lets us handle cases with NAs too.

Reply:
> bc ensure_data does the wrong thing with MaskedArrays, np.asarray(obj) defaults to object dtype.

this is very unfortunate. I don't really like this approach here. i suppose its ok for a backport though this is na experimental type and so i don't consider this regression to be a big deal.

Reply:
the more i work on it, the more i want to nuke NA from space

Reply:
ok what is involved with changing this to a non-recursive formulation though? e.g. ensure_data *should* be able to handle NA (if not we will have other issues)

Reply:
Three options:

1) have ensure_data special-case MaskedArray cases with no NAs, in which case they can just use the ndarray.  This fixes the regression (cases with NAs didnt use to work IIUC).  kluuuuudge
2) make algos.kth_smallest support object dtype (and not choke on pd.NA)
3) make the EA implement its own nlargest

I decided the approach here was less kludgy than those in part bc this function uses `obj.dropna()`, so the MaskedArray case actually is much simpler to implement than an arbitrary EA

Reply:
why is 1 a kludge?

Reply:
bc its special-casing for MaskedArray and special casing for `not values.isna().any()`

Reply:
worse than that, it isnt `not values.isna().any()` but `not values._mask.any()`